persist/fs/index_lookup.go

```diff
@@ -0,0 +1,197 @@
+		decoder:           decoder,
+		msgpackDecoder:    msgpack.NewDecoder(nil),
+	}
+}
+
+// getNearestIndexFileOffset returns either:
+//     1. The offset in the index file for the specified series
+//     2. The offset in the index file for the the series in the summaries file
+//        that satisfies the following two constraints:
+//            1. Is closest to the desired series in the index file
+//            2. Is BEFORE the desired series in the index file (because we
+//               we scan the index file sequentially in a forward-moving manner)
+// In other words, the returned offset can always be used as a starting point to
+// begin scanning the index file for the desired series.
+func (il *nearestIndexOffsetLookup) getNearestIndexFileOffset(id ts.ID) (int64, bool, error) {
+	idBytes := id.Data().Get()
+
+	min := 0
+	max := len(il.summaryIDsOffsets) - 1
+
+	// The summaries file only contains a fraction of the series that are in
+	// the index file itself. Because of that, the binary search that we're
+	// performing is "optimistic". We're trying to find either an exact match,
+	// OR the nearest match that is to the left of the series we're searching
+	// for (so we keep track of it everytime we move right).
+	bestMatchSoFar := int64(-1)
+
+	for {
+		if min > max {
+			if bestMatchSoFar == -1 {
+				return -1, false, nil
+			}
+			return bestMatchSoFar, true, nil
+		}
+
+		idx := (max + min) / 2
+		summaryBytesMetadata := il.summaryIDsOffsets[idx]
+		compBytes := summaryBytesMetadata.ID(il.summariesBytes)
+		comparison := bytes.Compare(idBytes, compBytes)
+
+		// Found it
+		if comparison == 0 {
+			indexOffset, err := summaryBytesMetadata.IndexOffset(
+				il.summariesBytes, il.decoderStream, il.msgpackDecoder)
+			// Should never happen, either something is really wrong with the code or
+			// the file on disk was corrupted
+			if err != nil {
+				return -1, false, err
+			}
+			return indexOffset, true, nil
+		}
+
+		// idBytes is smaller than compBytes, go left
+		if comparison == -1 {
+			max = idx - 1
+			continue
+		}
+
+		// idBytes is larger than compBytes, go right
+		if comparison == 1 {
+			min = idx + 1
+			indexOffset, err := summaryBytesMetadata.IndexOffset(
+				il.summariesBytes, il.decoderStream, il.msgpackDecoder)
+			if err != nil {
+				return -1, false, nil
+			}
+			// update the bestMatchSoFar everytime we move right
+			bestMatchSoFar = indexOffset
+			continue
+		}
+	}
+}
+
+func (il *nearestIndexOffsetLookup) close() error {
+	return munmap(il.summariesBytes)
+}
+
+// readNearestIndexOffsetLookupFromSummaries creates an nearestIndexOffsetLookup
+// from an index summaries file by reading the summaries file into an anonymous
+// mmap'd region, and also creating the slice of summaries offsets which is
+// required to binary search the data structure. It will also make sure that
+// the summaries file is sorted (which it always should be).
+func readNearestIndexOffsetLookupFromSummaries(
```

Comment:
nit: Might make sense to just prefix this with `new` since its really kind of a specialized constructor?

Reply:
sure